src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSyncController.scala

```diff
@@ -0,0 +1,171 @@
+package io.iohk.ethereum.blockchain.sync
+
+import akka.actor._
+import io.iohk.ethereum.blockchain.sync.FastSyncController.StartRegularSync
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.domain.{Block, BlockHeader, Blockchain}
+import io.iohk.ethereum.network.PeerActor.{BroadcastBlocks, MessageReceived, SendMessage, Subscribe}
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.{NewBlock, Status}
+import io.iohk.ethereum.network.p2p.messages.PV62._
+import io.iohk.ethereum.network.p2p.validators.BlockValidator.BlockError
+import io.iohk.ethereum.utils.Config
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.Try
+
+trait RegularSyncController {
+  self: Actor with ActorLogging with BlacklistSupport =>
+
+  val appStateStorage: AppStateStorage
+  val blockchain: Blockchain
+  val blockValidator: (BlockHeader, BlockBody) => Either[BlockError, Block]
+  def peersToDownloadFrom: Map[ActorRef, Status]
+
+  private var headers: Seq[BlockHeader] = Seq.empty
+  private var resolvingBranch = false
+  private var broadcasting = false
+
+  import Config.FastSync._
+
+  def regularSync(): Receive = {
+
+    case StartRegularSync =>
+      askForHeaders()
+
+    case ResumeRegularSync =>
+      askForHeaders()
+
+    case MessageReceived(m: BlockHeaders) if !resolvingBranch =>
+      handleDownload(m)
+
+    case MessageReceived(m: BlockHeaders) if resolvingBranch =>
+      handleBlockBranchResolution(m)
+
+    case MessageReceived(m: BlockBodies) =>
+      handleBlockBodies(m)
+
+    case m: BroadcastBlocks if broadcasting =>
+      peersToDownloadFrom.keys.foreach(_ ! m)
+
+    case ResolveBranch(peer) =>
+      requestHeadersForNewBranch(peer)
+  }
+
+  private def requestHeadersForNewBranch(peer: ActorRef) = {
+    resolvingBranch = true
+    peer ! SendMessage(GetBlockHeaders(Right(headers.head.parentHash), blockResolveDepth, skip = 0, reverse = true))
+  }
+
+  private def askForHeaders() = {
+
+    bestPeer match {
+      case Some(peer) =>
+        val blockNumber = appStateStorage.getBestBlockNumber()
+
+        peer ! Subscribe(Set(BlockHeaders.code, BlockBodies.code))
+        peer ! SendMessage(GetBlockHeaders(Left(blockNumber + 1), blockHeadersPerRequest, skip = 0, reverse = false))
```

Comment:
what happens if this peer does not send a response (for example, connections gets lost)? does regular sync stop in such situation?